lib/ex_aws/s3.ex

```diff
@@ -517,19 +517,21 @@ defmodule ExAws.S3 do
   |> ExAws.request! #=> :done
   ```
   """
-  def upload(source, bucket, path, opts \\ []) do
-    do_upload(source, bucket, path, opts)
-  end
-
-  defp do_upload(%Flow{} = flow, bucket, path, opts) do
+  @spec upload(
+    source :: Enumerable.t | Flow.t,
+    bucket :: String.t,
+    path :: String.t,
+    opts :: Keyword.t) :: __MODULE__.Upload.t
```

Comment:
I think this is fine. It might be nice however to document the opts, which would at this point just be `[max_concurrency: pos_integer]`